src/exchange/order_handler.cairo

```diff
@@ -177,50 +214,202 @@ mod OrderHandler {
             trigger_price: u128,
             min_output_amount: u128,
             order: Order
-        ) { // TODO
+        ) -> Order {
+            // Check only controller.
+            let role_module_state = RoleModule::unsafe_new_contract_state();
+            role_module_state.only_controller();
+
+            // Fetch data store.
+            let base_order_handler_state = BaseOrderHandler::unsafe_new_contract_state();
+            let data_store = base_order_handler_state.data_store.read();
+
+            global_reentrancy_guard::non_reentrant_before(data_store);
+
+            // Validate feature.
+            feature_utils::validate_feature(
+                data_store,
+                keys::update_order_feature_disabled_key(get_contract_address(), order.order_type)
+            );
+
+            assert(base_order_utils::is_market_order(order.order_type), 'OrderNotUpdatable');
+
+            let mut order = order.clone();
+            order.size_delta_usd = size_delta_usd;
+            order.trigger_price = trigger_price;
+            order.acceptable_price = acceptable_price;
+            order.min_output_amount = min_output_amount;
+            order.is_frozen = false;
+
+            // Allow topping up of execution fee as frozen orders will have execution fee reduced.
+            let wnt = token_utils::wnt(data_store);
+            let order_vault = base_order_handler_state.order_vault.read();
+            let received_wnt = order_vault.record_transfer_in(wnt).unwrap_syscall();
+            order.execution_fee += received_wnt;
+
+            let estimated_gas_limit = gas_utils::estimate_execute_order_gas_limit(
```

Comment:
same remark as for validate_execution_fee. It a struct must be passed and ownership kept -> ref if it's mutated, snapshot otherwise